pkg/util/awsutil/s3_client.go

```diff
@@ -0,0 +1,112 @@
+// Copyright 2017 The etcd-operator Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package awsutil
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+
+	api "github.com/coreos/etcd-operator/pkg/apis/etcd/v1beta2"
+
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	tmpdir = "/tmp"
+)
+
+// S3ClientFactory can be used to generate S3 clients from e.g. given aws secret.
+// It helps handle details around k8s logic and temp dir.
+type S3ClientFactory struct {
```

Comment:
maybe have a like struct that embeds `*s3.S3`
```go
type S3Client struct{
*s3.S3
configDir
}

func NewS3Client(kubecli kubernetes.Interface, namespace, awsSecret string)  *S3Client {
...
}

func(s *S3Client) Destroy(){
...
}
```

Reply:
Embedding won't  make *S3Client a *s3.S3.
Wrapping a S3Client is a good move.
Let's not worry about the embedding naming

Reply:
@hongchaodeng you are right. just realized that ` *s3.S3.` is not an interface. therefore Embedding won't work. I guess wrapper is the way to go then.